docs/reference/aggregations/pipeline/movfn-aggregation.asciidoc

```diff
@@ -0,0 +1,575 @@
+                 "doc_count": 3,
+                 "the_sum": {
+                    "value": 550.0
+                 }
+             },
+             {
+                 "key_as_string": "2015/02/01 00:00:00",
+                 "key": 1422748800000,
+                 "doc_count": 2,
+                 "the_sum": {
+                    "value": 60.0
+                 },
+                 "moving_fn": {
+                    "value": 550.0
+                 }
+             },
+             {
+                 "key_as_string": "2015/03/01 00:00:00",
+                 "key": 1425168000000,
+                 "doc_count": 2,
+                 "the_sum": {
+                    "value": 375.0
+                 },
+                 "moving_fn": {
+                    "value": 305.0
+                 }
+             }
+         ]
+      }
+   }
+}
+--------------------------------------------------
+// TESTRESPONSE[s/"took": 11/"took": $body.took/]
+// TESTRESPONSE[s/"_shards": \.\.\./"_shards": $body._shards/]
+// TESTRESPONSE[s/"hits": \.\.\./"hits": $body.hits/]
+
+
+==== Custom user scripting
+
+The Moving Function aggregation allows the user to specify any arbitrary script to define custom logic.  The script is invoked each time a
+new window of data is collected.  These values are provided to the script in the `values` variable.  The script should then perform some
+kind of calculation and emit a single `double` as the result.  Emitting `null` is not permitted, although `NaN` and +/- `Inf` are allowed.
+
+For example, this script will simply return the first value from the window:
+
+[source,js]
+--------------------------------------------------
+POST /_search
+{
+    "size": 0,
+    "aggs": {
+        "my_date_histo":{
+            "date_histogram":{
+                "field":"date",
+                "interval":"1M"
+            },
+            "aggs":{
+                "the_sum":{
+                    "sum":{ "field": "price" }
+                },
+                "the_movavg": {
+                    "moving_fn": {
+                        "buckets_path": "the_sum",
+                        "window": 10,
+                        "script": "return values[0];"
+                    }
+                }
+            }
+        }
+    }
+}
+--------------------------------------------------
+// CONSOLE
+// TEST[setup:sales]
+
+==== Pre-built Functions
+
+For convenience, a number of functions have been prebuilt and are available inside the `moving_fn` script context:
+
+- `windowMax()`
+- `windowMin()`
+- `windowSum()`
+- `simpleMovAvg()`
+- `linearMovAvg()`
+- `ewmaMovAvg()`
+- `holtMovAvg()`
+- `holtWintersMovAvg()`
+
+The functions are available from the `MovingFunctions` namespace.  E.g. `MovingFunctions.windowMax()`
+
+===== windowMax Function
+
+This function accepts a collection of doubles and returns the maximum value in that window.  If the window is empty, or all values
+are `null`/`NaN`, `NaN` is returned as the result.
```

Comment:
what happens if some of the values are `NaN`? does it ignore those `NaN` values and return the max of the others or return `NaN`?

Reply:
This also applies to the other functions

Reply:
Yes, it'll find the `<function>` of the non-`NaN` values, ignoring the `NaN`s.  I reasoned that you can't find the min/max/whatever of a `NaN` because it isn't a number, so there's no way to compare it to actual numbers and they should just be ignored.

That makes the averages a bit tricky though, since they are calculating the average based on number of non-`NaN` values not "positions in the window".  It feels like if the user wants to count-but-not-use `NaN`s, they should probably write their own avg function at that point.

I'll beef up the documentation to make this point clear, since it's not right now.

Note: `+/- Inf` are treated as proper numbers for all functions, since the you can say `1 < +Inf`